misk-jdbc/src/main/kotlin/misk/jdbc/DataSourceService.kt

```diff
@@ -113,6 +116,27 @@ class DataSourceService @JvmOverloads constructor(
       hikariConfig.dataSourceProperties["characterEncoding"] = "UTF-8"
     }
 
+    // TODO(sahilm): The same mitigation _might_ be applicable to the DataSourceTypes VITESS_MYSQL and TIDB
+    if (config.type == DataSourceType.MYSQL && config.mysql_enforce_writable_connections) {
+      /*
+          Q. Why isn't this a DataSourceDecorator?
+          A. Because HikariCP calls Connection.isValid() in its internal getConnection() method before returning the connection to the DataSourceDecorator.
+             Implementing this logic in a DataSourceDecorator is too late and thus must be wrapped in the DataSource that HikariCP uses internally.
+          Q. Why can't we use a connectionTestQuery like `SET [SESSION] TRANSACTION READ WRITE` as [proposed](https://groups.google.com/g/hikari-cp/c/VH7nqwGimCs) in the HikariCP mailing list?
+          A. Because MySQL does not raise an error if `SET [SESSION] TRANSACTION READ WRITE` is executed on a read only connection.
+
+          TODO(sahilm): Extract ConnectionDecoratingDataSource and WritableConnectionValidator to Wisp so the same mitigation can be used in Armeria,
+           should wait for the solution to prove itself first.
+       */
+      val mysqlDataSource = buildDataSource(hikariConfig)
+      hikariConfig.dataSource = ConnectionDecoratingDataSource(
+        connectionDecorator = { connection ->
```

Comment:
I might have another use case for this decorator. I might convert this into a multi binding so service owners can extend this if they want.